stl/inc/atomic

```diff
@@ -351,6 +366,25 @@ struct _Atomic_storage {
         return _Result;
     }
 
+    void wait(_Ty _Expected, const memory_order _Order = memory_order_seq_cst) const noexcept {
+        long _Spin_context = 0;
+        for (;;) {
+            _Ty _Observed = load(_Order);
+            if (_CSTD memcmp(&_Observed, _STD addressof(_Expected), sizeof(_Ty)) != 0)
+                break;
+            _Atomic_wait_indirect(_STD addressof(_Storage), _Spin_context);

```

Comment:
Non lock free atomic contains a spin mutex. So it is possible layout (although not the current one), that addressof(_Storage) != this. 

Sure it does not matter much for `_indirect` functions, since address is only a key for them. And for `_direct` there's no spin mutex. So, with `this` it all should work.

I think it would be cleaner just to use addressof(_Storage), to avoid relying on the proof above.

Reply:
True, but I think anyone changing the layout would have to be aware of what they are doing if they do that. (I'm trying to avoid needing to pay for the extra function call in debug mode where we don't have to)

Reply:
It is pre-existent in [compate_exchange_strong](https://github.com/microsoft/STL/blob/285187b7b24be4bebfa271fbd0b1eacd995d18f4/stl/inc/atomic#L338)